src/main/docs/man/man3/Intake.3
.TH "Intake" 3 "Thu Oct 31 2019" "2020 Template Project" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Intake
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Intake\&.h>\fP
.PP
Inherits \fBIMechanism\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBIntake\fP ()"
.br
.ti -1c
.RI "virtual \fB~Intake\fP ()"
.br
.ti -1c
.RI "\fBIMechanism::MECHANISM_TYPE\fP \fBGetType\fP () const override"
.br
.ti -1c
.RI "void \fBSetPercentOutput\fP (double value) override"
.br
.ti -1c
.RI "void \fBSetPosition\fP (double pos) override"
.br
.ti -1c
.RI "double \fBGetCurrentPosition\fP () const override"
.br
.ti -1c
.RI "double \fBGetTargetPosition\fP () const override"
.br
.ti -1c
.RI "void \fBSetSpeed\fP (double speed) override"
.br
.ti -1c
.RI "double \fBGetCurrentSpeed\fP () const override"
.br
.ti -1c
.RI "double \fBGetTargetSpeed\fP () const override"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
==================================================================================================================================================== Copyright 2019 Lake Orion Robotics FIRST Team 302
.PP
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
.PP
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software\&.
.PP
THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT\&. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE 
.SH "OR OTHER DEALINGS IN THE SOFTWARE\&."
.PP
======================================================================================================== 
.SH "Intake\&.h"
.PP
File Description: This is the \fBIntake\fP Subsystem
.PP
.PP
 ======================================================================================================== Class: \fBIMechanism\fP 
.SH "Description:   Interface for subsystems"
.PP

.PP
Definition at line 43 of file Intake\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Intake::Intake ()"
================================================================================== method: \fBIntake\fP <<constructor>> 
.SH "description:    Create the subobjects and initialize the Intake subsystem"
.PP
==================================================================================================================================================== Copyright 2019 Lake Orion Robotics FIRST Team 302
.PP
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
.PP
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software\&.
.PP
THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT\&. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE 
.SH "OR OTHER DEALINGS IN THE SOFTWARE\&."
.PP
======================================================================================================== 
.SH "Intake\&.cpp"
.PP
File Description: This is the \fBIntake\fP Subsystem
.PP
.PP
 ================================================================================== method: \fBIntake\fP <<constructor>> 
.SH "description:    Create the subobjects and initialize the Intake subsystem"
.PP

.PP
Definition at line 47 of file Intake\&.cpp\&.
.SS "Intake::~Intake ()\fC [virtual]\fP"
================================================================================== method: ~Intake <<destructor>> 
.SH "description:    clean up memory when this object gets deleted"
.PP
================================================================================== method: ~Intake <<destructor>> 
.SH "description:    clean up memory when this object gets deleted"
.PP

.PP
Definition at line 66 of file Intake\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "double Intake::GetCurrentPosition () const\fC [override]\fP, \fC [virtual]\fP"
================================================================================== method: GetCurrentPostion description: Return the current position of the intake in degrees\&. Since we don't have a sensor this will return -90 for clockwise rotations and 90 for counter-clockwise rotations\&. 
.SH "returns:        double  position in degrees (rotating mechansim)"
.PP
================================================================================== method: GetCurrentPostion description: Return the current position of the intake in degrees\&. Since we don't have a sensor this will return -90 for clockwise rotations and 90 for counter-clockwise rotations\&. 
.SH "returns:        double  position in degrees (rotating mechansim)"
.PP

.PP
Implements \fBIMechanism\fP\&.
.PP
Definition at line 139 of file Intake\&.cpp\&.
.SS "double Intake::GetCurrentSpeed () const\fC [override]\fP, \fC [virtual]\fP"
================================================================================== method: GetCurrentSpeed description: Get the current speed of the intake\&. The value is in degrees per second\&. Since we don't have a sensor for this mechanism, it will return -360 for clockwise rotations and 360 for counter clockwise rotations\&. 
.SH "returns:        double  speed in degrees/second (rotating mechansim)"
.PP
================================================================================== method: GetCurrentSpeed description: Get the current speed of the intake\&. The value is in degrees per second\&. Since we don't have a sensor for this mechanism, it will return -360 for clockwise rotations and 360 for counter clockwise rotations\&. 
.SH "returns:        double  speed in degrees/second (rotating mechansim)"
.PP

.PP
Implements \fBIMechanism\fP\&.
.PP
Definition at line 199 of file Intake\&.cpp\&.
.SS "double Intake::GetTargetPosition () const\fC [override]\fP, \fC [virtual]\fP"
================================================================================== method: GetTargetPostion description: Return the target position of the intake\&. Since we don't have a sensor this will return -90 for clockwise rotations and 90 for counter-clockwise rotations\&. 
.SH "returns:        double  position in degrees (rotating mechansim)"
.PP
================================================================================== method: GetTargetPostion description: Return the target position of the intake\&. Since we don't have a sensor this will return -90 for clockwise rotations and 90 for counter-clockwise rotations\&. 
.SH "returns:        double  position in degrees (rotating mechansim)"
.PP

.PP
Implements \fBIMechanism\fP\&.
.PP
Definition at line 155 of file Intake\&.cpp\&.
.SS "double Intake::GetTargetSpeed () const\fC [override]\fP, \fC [virtual]\fP"
================================================================================== method: GetTargetSpeecd description: Get the target speed of the intake\&. The value is in degrees per second\&. Since we don't have a sensor for this mechanism, it will return -360 for clockwise rotations and 360 for counter clockwise rotations\&. 
.SH "returns:        double  speed in degrees/second (rotating mechansim)"
.PP
================================================================================== method: GetTargetSpeecd description: Get the target speed of the intake\&. The value is in degrees per second\&. Since we don't have a sensor for this mechanism, it will return -360 for clockwise rotations and 360 for counter clockwise rotations\&. 
.SH "returns:        double  speed in degrees/second (rotating mechansim)"
.PP

.PP
Implements \fBIMechanism\fP\&.
.PP
Definition at line 216 of file Intake\&.cpp\&.
.SS "\fBIMechanism::MECHANISM_TYPE\fP Intake::GetType () const\fC [override]\fP, \fC [virtual]\fP"
================================================================================== method: GetType description: Indicates this is the intake 
.SH "returns:        IMechanism::MECHANISM_TYPE::INTAKE"
.PP
================================================================================== method: GetType description: Indicates this is the intake 
.SH "returns:        IMechanism::MECHANISM_TYPE::INTAKE"
.PP

.PP
Implements \fBIMechanism\fP\&.
.PP
Definition at line 76 of file Intake\&.cpp\&.
.SS "void Intake::SetPercentOutput (double value)\fC [override]\fP, \fC [virtual]\fP"
================================================================================== method: SetPercentOutput description: Run intake in open loop (percent output) 
.SH "returns:        void"
.PP
================================================================================== method: SetPercentOutput description: Run intake in open loop (percent output) 
.SH "returns:        void"
.PP

.PP
\fBParameters\fP
.RS 4
\fIvalue\fP \fI - percent output for the motor(s) \fP
.RE
.PP

.PP
Implements \fBIMechanism\fP\&.
.PP
Definition at line 88 of file Intake\&.cpp\&.
.SS "void Intake::SetPosition (double pos)\fC [override]\fP, \fC [virtual]\fP"
================================================================================== method: SetPosition description: Run intake in closed loop position mode\&. The value is in degrees\&. Since we don't have a sensor, this will run percent output in the direction specified\&. We will use the standard unit circle for direction with negative being clockwise and positive being counter-clockwise\&. 
.SH "returns:        void"
.PP
================================================================================== method: SetPosition description: Run intake in closed loop position mode\&. The value is in degrees\&. Since we don't have a sensor, this will run percent output in the direction specified\&. We will use the standard unit circle for direction with negative being clockwise and positive being counter-clockwise\&. 
.SH "returns:        void"
.PP

.PP
\fBParameters\fP
.RS 4
\fIpos\fP \fI - target position in degrees (rotating mechansim) \fP
.RE
.PP

.PP
Implements \fBIMechanism\fP\&.
.PP
Definition at line 113 of file Intake\&.cpp\&.
.SS "void Intake::SetSpeed (double speed)\fC [override]\fP, \fC [virtual]\fP"
================================================================================== method: SetSpeed description: Run intake in closed loop velocity mode\&. The value is in degrees/second (rotating mechansim)\&. Since we don't have a sensor for this mechanism, it will return -360 for clockwise rotations and 360 for counter clockwise rotations\&. 
.SH "returns:        void"
.PP
================================================================================== method: SetSpeed description: Run intake in closed loop velocity mode\&. The value is in degrees/second (rotating mechansim)\&. Since we don't have a sensor for this mechanism, it will return -360 for clockwise rotations and 360 for counter clockwise rotations\&. 
.SH "returns:        void"
.PP

.PP
\fBParameters\fP
.RS 4
\fIspeed\fP \fI - target speed degrees/second (rotating mechansim) \fP
.RE
.PP

.PP
Implements \fBIMechanism\fP\&.
.PP
Definition at line 171 of file Intake\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for 2020 Template Project from the source code\&.


src/main/cpp/subsys/Intake.cpp

///====================================================================================================================================================
/// Copyright 2019 Lake Orion Robotics FIRST Team 302
///
/// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"),
/// to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
/// and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
///
/// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
///
/// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
/// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
/// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
/// OR OTHER DEALINGS IN THE SOFTWARE.
///====================================================================================================================================================

///========================================================================================================
/// Intake.cpp
///========================================================================================================
///
/// File Description:
///     This is the Intake Subsystem
///
///========================================================================================================

// C++ Includes
#include <algorithm>
#include <cmath>
#include <iostream>

// FRC includes

// Team 302 includes
#include <hw/DragonMotorControllerFactory.h>        
#include <hw/IDragonMotorController.h>
#include <subsys/IMechanism.h>
#include <subsys/Intake.h>

// Third Party Includes



///==================================================================================
/// method:         Intake <<constructor>>
/// description:    Create the subobjects and initialize the Intake subsystem
///==================================================================================
Intake::Intake()
{
    // Get the motor controller and set its mode to percent output and stop it
    /**
    m_motor = DragonMotorControllerFactory::GetInstance()->GetController( IMechanism::MECHANISM_TYPE::INTAKE, IDragonMotorController::MOTOR_CONTROLLER_TYPE::INTAKE );
    if ( m_motor != nullptr )
    {
        m_motor->SetControlMode( IDragonMotorController::DRAGON_CONTROL_MODE::PERCENT_OUTPUT );
        m_motor->Set( 0.0 );
    }
    else
    {
        std::cout << "==>>Intake::Intake motor not found " << std::endl;
    }
    **/
}

///==================================================================================
/// method:         ~Intake <<destructor>>
/// description:    clean up memory when this object gets deleted
///==================================================================================
Intake::~Intake()
{
    // delete owned objects (e.g. delete m_motor;)
}

///==================================================================================
/// method:         GetType
/// description:    Indicates this is the intake
/// returns:        IMechanism::MECHANISM_TYPE::INTAKE
///==================================================================================
IMechanism::MECHANISM_TYPE Intake::GetType() const 
{
    return IMechanism::MECHANISM_TYPE::INTAKE;
}


///==================================================================================
/// method:         SetPercentOutput
/// description:    Run intake in open loop (percent output)
/// returns:        void
///==================================================================================
void Intake::SetPercentOutput
(
    double      value       /// <I> - percent output for the motor(s)
) 
{
    // Make sure value is in range (-1.0 to 1.0) and then set the percent output
    // on the motor (2 calls)
    if ( m_motor != nullptr )
    {
        auto pct = std::min(1.0, std::max(value, -1.0));  // replace std::clamp in c++17
        m_motor->SetControlMode( IDragonMotorController::DRAGON_CONTROL_MODE::PERCENT_OUTPUT );
        m_motor->Set( pct );
    }
}


///==================================================================================
/// method:         SetPosition
/// description:    Run intake in closed loop position mode.  The value is in 
///                 degrees.  Since we don't have a sensor, this will run percent
///                 output in the direction specified.  We will use the standard unit
///                 circle for direction with negative being clockwise and positive
///                 being counter-clockwise.
/// returns:        void
///==================================================================================
void Intake::SetPosition 
(
    double      pos       /// <I> - target position in degrees (rotating mechansim) 
) 
{
    // Since there isn't a sensor, just set the motor speeds based on the position
    // input to either -1.0 or 1.0. Just call the set method with these values.
    double percent = 0.0;
    if ( pos < m_deadbandTol )
    {
        percent = 1.0;
    }
    else if ( pos > m_deadbandTol )
    {
        percent = -1.0;
    }
    SetPercentOutput( percent );
}


///==================================================================================
/// method:         GetCurrentPostion
/// description:    Return the current position of the intake in degrees.  Since
///                 we don't have a sensor this will return -90 for clockwise rotations
///                 and 90 for counter-clockwise rotations.
/// returns:        double  position in degrees (rotating mechansim)
///==================================================================================
double Intake::GetCurrentPosition() const 
{
    // Normally would call GetSelectedSensorPosition, but there is no sensor, so we
    // would call GetMotorOutputPercent.  This isn't currently exposed in our APIs, so 
    // hard code for now 
    return 90.0;
}


///==================================================================================
/// method:         GetTargetPostion
/// description:    Return the target position of the intake.  Since
///                 we don't have a sensor this will return -90 for clockwise rotations
///                 and 90 for counter-clockwise rotations.
/// returns:        double  position in degrees (rotating mechansim)
///==================================================================================
double Intake::GetTargetPosition() const 
{
    // No sensor so it is the same as the current
    return GetCurrentPosition();
}


///==================================================================================
/// method:         SetSpeed
/// description:    Run intake in closed loop velocity mode.  The value is in 
///                 degrees/second (rotating mechansim).  Since we don't have a 
///                 sensor for this mechanism, it will return -360 for clockwise 
///                 rotations and 360 for counter clockwise rotations.
/// returns:        void
///==================================================================================
void Intake::SetSpeed 
(
    double      speed       /// <I> - target speed degrees/second (rotating mechansim)
) 
{
    // Since there isn't a sensor, just set the motor speeds based on the position
    // input to either -1.0 or 1.0. Just call the set method with these values.
    // TODO:: maybe we should run a voltqge control mode instead
    double percent = 0.0;
    if ( speed < m_deadbandTol )
    {
        percent = 1.0;
    }
    else if ( speed > m_deadbandTol )
    {
        percent = -1.0;
    }
    SetPercentOutput( percent );
}


///==================================================================================
/// method:         GetCurrentSpeed
/// description:    Get the current speed of the intake.  The value is in degrees 
///                 per second.  Since we don't have a sensor for this mechanism,
///                 it will return -360 for clockwise rotations and 360 for 
///                 counter clockwise rotations.
/// returns:        double  speed in degrees/second (rotating mechansim)
///==================================================================================
double Intake::GetCurrentSpeed() const 
{
    // Normally would call GetSelectedSensorVelocity, but there is no sensor, so we
    // would call GetMotorOutputPercent.  This isn't currently exposed in our APIs, so 
    // hard code for now 
    return 360.0;
}


///==================================================================================
/// method:         GetTargetSpeecd
/// description:    Get the target speed of the intake.  The value is in degrees 
///                 per second.  Since we don't have a sensor for this mechanism,
///                 it will return -360 for clockwise rotations and 360 for 
///                 counter clockwise rotations.
/// returns:        double  speed in degrees/second (rotating mechansim)
///==================================================================================
double Intake::GetTargetSpeed() const 
{
    // No sensor so it is the same as the current
    return GetCurrentPosition();
}





